src/db/models/user.model.js

```diff
@@ -129,11 +142,90 @@ module.exports = (sequelize, DataTypes) => {
+        // This expects and email, and an expiration time window for storing a
+        // messaging protocol `messagingProtocol` token, auth expiration, and
+        // provider for a user. Users will be asked to validate this token against
+        // the submission of `verifyMessagingProtcolToken` (With or without
+        // credentials)
+
+        // QUESTION: Do we want to add some form of rate limiting in this flow?
+        return this.find({
+            where: {
+                email,
+            },
+        })
+        .then((user) => {
+            if (!user) {
+                const err = new Error('Email not found');
+                return sequelize.Promise.reject(err);
+            }
+            // JRB removed protocol, because we can confirm this. Put back it leads to errors.
+            // JRB: This is new
+            return user.updateVerifyAccountToken(email, expTime);
+        });
+    };
+
+    User.getVerifyingUser = function getVerifyingUser(token) {
+        // This expects a `contactMethodVerificationToken`, and provides a user's username.
+        return this.findOne({
+            where: {
+                contactMethodVerificationToken: token,
+            },
+        })
+        .then((user) => {
+            if (!user) {
+                const err = new Error('Token not found');
+                return sequelize.Promise.reject(err);
+            }
+            return user.username;
+        });
+    };
+
+    User.verifyUserAccount = function verifyUserAccount(token) {
+        return this.find({
+            where: {
+                contactMethodVerificationToken: token,
+            },
+        })
+        .then((user) => {
+            if (!user) {
+                const err = new Error('Token not found');
+                return sequelize.Promise.reject(err);
+            }
+            return user.addVerifiedContact(user.contactMethodToVerify);
+        });
+    };
+
+    User.secureVerifyUserAccount = function secureVerifyUserAccount(token, password) {
+        return this.find({
+            where: {
+                contactMethodVerificationToken: token,
+            },
+        })
+        .then((user) => {
+            if (!user) {
+                const err = new Error('Token not found');
+                return sequelize.Promise.reject(err);
+            }
+            const verificationPassword = crypto.pbkdf2Sync(password, Buffer.from(user.salt), 100000, 128, 'sha256').toString('hex');
+            if (user.password === verificationPassword) {
+                return user.addVerifiedContact(user.contactMethodToVerify);
+            }
+            const err = new Error('Password does not match.');
+            return sequelize.Promise.reject(err);
+        });
+    };
+
     // Instance methods
     User.prototype.isAdmin = function isAdmin() {
         return this.scopes.includes('admin');
     };
 
+    User.prototype.isVerified = function isVerified() {
```

Comment:
This will not work if we try to expand to verifying other contact methods. Not sure if you want to make this function support multiple contact methods right now, or not.


I would think the solution would be
```js
// Here in this file
User.prototype.isVerified = function isVerified(contactMethod) {
    return this.verifiedContactMethods.includes(contactMethod);
}
// To call it from elsewhere in the code
myUser.isVerified(myUser.email); // or myUser.phone, or something even more dynamic than that
```

Reply:
I'm fine either way. Using `isVerified()` doesn't prevent extension in the future, just means we're stuck with email as the default contact method if we want back-compatibility. Email as default isn't a bad default to my mind.

Reply:
Not following this comment

Reply:
Jon and I discussed. Resolution: No action, because…
- The class method takes an email arg and looks up the one user is belongs to; and if the user exists, it calls the instance method
- The instance method generates a verification token for this user
They are broken apart because we might want to call the token generation instance method on an array of users.

Reply:
Resolution: We are leaving it this way for now because, if we want to change it later, that won't be terribly hard to do.